### ISSUE TITLE
Retry is not working when there is an error "TimeoutError: retry timed out"

### ISSUE THREAD

**Issue by Kushi171987** [mickhansen/retry-as-promised#23]
I used "the retry-as-promised" in my Sequelize node js server to connect the database.
I added match as 
`match: [ 
                Sequelize.ConnectionError,
                Sequelize.ConnectionRefusedError,
                Sequelize.ConnectionTimedOutError,
                Sequelize.OptimisticLockError,
                Sequelize.TimeoutError,
                "TimeoutError",
                "retryTimeout",
                "TimeoutError: retry timed out",
                // 'SequelizeDatabaseError: Deadlock found when trying to get lock; try restarting transaction',
                // 'ER_LOCK_DEADLOCK',
                // 'SQL_BUSY',
                /SequelizeConnectionError/,
                /SequelizeConnectionRefusedError/,
                /SequelizeConnectionTimedOutError/,
                /SequelizeHostNotFoundError/,
                /SequelizeHostNotReachableError/,
                /SequelizeInvalidConnectionError/,
                /TimeoutError/
            ]`

though I added the "TimeoutError", its not retrying when happened with that error.

**Comment by mickhansen** [comment]
The timeout error generated by the library is outside the retry logic loop: https://github.com/mickhansen/retry-as-promised/blob/master/index.js#L66

**Comment by Kushi171987** [comment]
Thank you for the quick response.

 I got it now. 
To fix the problem I have an option to increase the value of timeout param, but query execution time depends on network traffic & server load. 

So is there any option to retry that kind of error?

**Comment by mickhansen** [comment]
Not at this time, needs to be fixed.

**Comment by blipk** [comment]
> The timeout error generated by the library is outside the retry logic loop: https://github.com/mickhansen/retry-as-promised/blob/master/index.js#L66

@mickhansen 

Does this not completely break the purpose?
If the timeout logic is outside of the retry logic then there will never be a retry if the timeout is hit?

Am I missing something?

Attempting to use this in sequelize and this fails without retrying:
```ts
    const retryOptions: RetryOptions = {
        match           : [ TimeoutError ],
        report          : console.log,
        max             : 5,
        timeout         : 1 * 1000, // 1 second timeout
        backoffBase     : 10 * 1000,
        backoffExponent : 1
    }
    const options: QueryOptions = { retry: retryOptions }

//takes ~5 seconds on my CPU
    await sequelize.queryRaw( `WITH RECURSIVE r(i) AS (
  VALUES(0)
  UNION ALL
  SELECT i FROM r
  LIMIT 10000000
)
SELECT i FROM r WHERE i = 1;`, options )
```

**Comment by mickhansen** [comment]
@blipk Correct, as the code is written now, if the timeout is hit it does not trigger retries. I can't recall if that was intended or not, but in retrospect it does not sound ideal.

**Comment by blipk** [comment]
Thanks for the quick response and confirmation @mickhansen 

Are you intending to work on this?

Would you accept a PR? Any preference on how you would want it implemented?

**Comment by mickhansen** [comment]
> Thanks for the quick response and confirmation @mickhansen
> 
> Are you intending to work on this?
> 
> Would you accept a PR? Any preference on how you would want it implemented?

@blipk I imagine it would be as simple as converting the current setup to use `Promise.race` instead. The code is generally a bit ancient at this point, I will happily accept PRs - but better libraries may exist :)

**Comment by blipk** [comment]
Thanks @mickhansen 

My concern is that Sequelize uses this library for its underlying retry logic

I've raised an issue there and will see what they say - https://github.com/sequelize/sequelize/issues/17459
